### ISSUE TITLE
Lists and Song APIs to be added.

### ISSUE THREAD

**Issue by eliasalbro95** [thomasdilts/churchsongs#86]
### We need some new APIs to enhance our functionality:

**Set APIs**
1. API to modify/edit specific elements(key, notes, verses/choruses) of a set, such as a song or a comment.
2. API to change the order of an element in a set (I’m not sure if this can be done as an API or if we need to pass the new order to the existing list change API).

**Song APIs**
1. Malek may have mentioned this, but we added a feature for users to create a new song by selecting specific verses/choruses from existing songs and combining them into a medley. For this, I will use the existing song add API, but I need it to return the full information of the added song, instead of calling another API (song get) to retrieve it. I know that song add currently returns only the song’s id, but it would be helpful if we can get the whole song object.
2. Malek was discussing adding an edit song request by the users, lyrics, and chords, so I was wondering, where should I submit the changes? Song change API or do we need to specify a new one called song change request API?

**Comment by thomasdilts** [comment]
I agree with the "Set APIs" and will do them as soon as possible.
The song APIs:
1. This is totally new information for me. This would be rather difficult to do. I think we must put this on at a much later date. We need to focus on getting a releasable system first. Crazy ideas like this can be added later.
2. This I am aware of. I already have a system done for just this exact purpose both frontend and backend. However, Malek doesn't like at all how it looks on the frontend. I agree, it needs to be made MUCH simpler. Still, it's only a matter of getting the GUI to be more user friendly. Thats all. The system is finished and working.

**Comment by thomasdilts** [comment]
For the edit Song API  backend it is simply the call to 

https://github.com/thomasdilts/churchsongs/wiki/Song-change

Thats it. No more is needed on the backend. The frontend however needs much work to be simpler..

**Comment by eliasalbro95** [comment]
I start from bottom to top:

1. Yes that's the point we need to make it as simple as we can, it gonna need time to be implemented.

2. I have 2 points:
**Usefulness:** I think the Medley feature would be very useful for worship leaders like me, our teams, and many others. It would allow them to create and play medleys of songs with the verses they want in the order they want. We already have many great features, but this one would make worship leading time even more simpler and creative. I know many worship leaders who love to make medleys in their worship sessions, but sometimes it's hard for the instrument players they serve with, they need to have the notes in front of them and know exactly what the keys(chords) are, so they can't do what they have to because it's hard to make everyone knows what is the chords, so this would make it doable even if we have humble instruments players.
**Complexity:**
I think it's simple to implement at least from my point of view(correct me if I'm wrong).
Simply we add a new song with Song Add API using the selected verses/choruses and add it only to the user's set, plus we need to add it to the submitted songs so the admins can check and maybe publish it under the medley tag.
let me know if I'm missing something.
_**Note:**_ and I'm already implementing the frontend process for that, so I only need to be sure that the added song will be passed with the response body not only the id.

**Comment by thomasdilts** [comment]
Ok. I see it could be a good thing to have. But the complexity is the same as the whole "set" system that we have and I think the "set" system is very complex. If we don't focus on getting delivered the basic functionality then I am worried that zamar will never get finished. Lets do this but only after zamar is released to the public.

**Comment by eliasalbro95** [comment]
I understand what you say, after all, we can't add every idea we have now, we have to stop and release at some point.

But I only need the song's information to complete it in the application, I implemented the feature, and for saving time and tracking purposes I asked you to return the song's body and not to call the Song Get API every time the user create a medley.
So if you only do this and the rest we can do it later. 

Thank you and God bless.

**Comment by eliasalbro95** [comment]
**Firstly** I think we are mot talking about the exact same thing, discussing an idea with only writing has many difficulties and this is one of them 😅.

About what you said, yes, that's a kind of the same thing, but what we are talking about here is a feature that allows me as a user to add a song has many verses/choruses of other existed songs, in the same song model ( I mean the same page) and it will be as an item of a set, not several songs after each other has only the verses/choruses the user chose.


About the **complexity**:
I'm gonna say what I'm thinking about it and let me know if that's doable or not from backend prospective.

**Frontend:** The user will submit a new song using an existing API (song add) and expecting the to be add to the admin's song list that need to be approved, **but** anyways it will be added to the users set wether the admin approves it or not.

Hope this make it clearer. 
Let me know what you think.

**Comment by eliasalbro95** [comment]
And let me add the following, actually, we can implement the feature in our current APIs, by doing the following:

1. add a new song by the user (song add API).
2. get the song info (song get API)
3. then add the song to the set.

but simply what I'm trying to say and asking you if we can make is making this process one API call instead of 3 calls.

**Comment by thomasdilts** [comment]
ok. Yes. I understand. Such a change would of course be very easy for me to do for the backend It would take about an hour. But if I would try to do that in the GUI-frontend it would take me a very very long time. I need to discuss this with malek because he was at one point against allowing the user to create songs. Maybe we could give these types of songs a special status-- like MEDLEY instead of SUBMITTED which is the status they get today.

**Comment by MalekBakary** [comment]
It is a good idea to give it special status (medley as you suggested). Because in this case they are not submitting something new that we need to review and check if it is suitable for our terms, the just make a special order of songs that we already have in our library. My suggestion is to give them the ability to even share this new created medley with their friends (cuz it's in a set). And give the admin the ability to add it to our song library (to make it visible for the public). What do you think? Is it possible from backend prospective? 

When it comes to GUI (which is the website, I assume😅) I think you don't have to make it right now. We can release this feature on the app and we can add it to the website in the future.

**Comment by thomasdilts** [comment]
The requested list apis are now finished.
https://github.com/thomasdilts/churchsongs/wiki/List-changesong
https://github.com/thomasdilts/churchsongs/wiki/List-sequence

**Comment by thomasdilts** [comment]
The requested medley api is now finished. Good luck implementing the gui. We really need to freeze the functionality of zamar otherwise we will never be finished.
https://github.com/thomasdilts/churchsongs/wiki/Song-addmedley
https://github.com/thomasdilts/churchsongs/wiki/Song-changemedley
https://github.com/thomasdilts/churchsongs/wiki/Song-deletemedley

I am now closing this issue. This issue had WAY too many items addressed in the issue. You really need to make the issues smaller so that they are more readable. If there is something I missed in the issue then open a new issue. Don't reopen this issue.
